File: ui/composite/TrackFrame.py
__author__ = 'toramisu'
from ui import *
from utils import B
from module.Events import *


class TrackFrame(QWidget):
    def __init__(self, parent):
        super(TrackFrame, self).__init__(parent)
        self.thumbPixmap = None
        self.relayout = None
        self.changeWidth = 0
        self.resize(40, 60)

        self.thumb = QWidget(self)
        self.thumb.resize(40, 50)
        self.thumb.move(0, 0)
        setMouseTransparent(self.thumb)
        self.thumb.setStyleSheet(
            'border:none'
        )
        self.thumb.paintEvent = self.thumbPaintEvent
        self.holdFrameCount = 1
        self.holdFrameCountLabel = None
        self.frameIdx = QLabel(self)
        self.frameIdx.setStyleSheet('color:#b6b6b6;border:none;')
        setMouseTransparent(self.frameIdx)
        self.frameIdx.move(4, 45)
        self.frameIdx.setText('0')
        self.leftButton = QPushButton(self)
        self.isPressLeftButton = False
        self.leftButton.resize(8, 8)
        setStyle(self.leftButton, ':qss_button', 'frameButton')
        self.leftButton.mousePressEvent = self.onPressLeftButton
        self.leftButton.mouseReleaseEvent = self.onRels

        self.rightButton = QPushButton(self)
        self.isPressRightButton = False
        setStyle(self.rightButton, ':qss_button', 'frameButton')
        self.rightButton.resize(8, 8)
        self.rightButton.move(33, 32)
        self.rightButton.mousePressEvent = self.onPressRightButton
        self.rightButton.mouseReleaseEvent = self.onRels
        self.mousePressEvent = self.onPress
        self.mouseMoveEvent = self.onMove
        self.mouseReleaseEvent = self.onRels
        self.isPress = False
        self.__refSImage = None

    @property
    def refSImage(self):
        return self.__refSImage

    @refSImage.setter
    def refSImage(self, value):
        self.__refSImage = value

    def onRels(self, e):
        self.isPress = False
        self.isPressRightButton = False
        if self.isPressLeftButton:
            Event.dis(ActionEvent.RELEASE_TRACK_FRAME_LEFT_BUTTON)
            self.isPressLeftButton = False
        pass

    def onMove(self, e):
        if self.isPress:
            posX = e.pos().x()
            if self.isPressRightButton:
                if posX > self.width() + 30:
                    self.setHoldFrameCount(self.holdFrameCount + 1)
                    self.relayout(self)
                    print(self, self.width(), e.pos().x())
                elif posX < self.width() - 30 and self.holdFrameCount > 1:
                    self.setHoldFrameCount(self.holdFrameCount - 1)
                    self.relayout(self)
            elif self.isPressLeftButton:
                if posX < -30:
                    self.setHoldFrameCount(self.holdFrameCount + 1, -1)
                    self.relayout(self)
                    pass
                elif posX > 30 and self.holdFrameCount > 1:
                    self.setHoldFrameCount(self.holdFrameCount - 1, -1)
                    self.relayout(self)
                    pass
            else:
                self.parent().mouseMoveEvent(e)
            pass
        pass

    def onPress(self, e):
        self.isPress = True
        self.parent().mousePressEvent(e)

    def onPressRightButton(self, e):
        self.isPress = True
        self.isPressRightButton = True

    def onPressLeftButton(self, e):
        self.isPress = True
        self.isPressLeftButton = True

    def getIdx(self):
        return int(self.frameIdx.text().replace('.', ''))

    def setIdx(self, idx):
        self.frameIdx.setText(str(idx) + '.')

    def setPixmap(self, img):
        self.thumbWidth = img.width()
        self.thumbHeight = img.height() / img.width() * (self.width() - 3)
        self.thumbPixmap = img.scaled(self.width(), self.thumbHeight)
        pass

    def thumbPaintEvent(self, QPaintEvent):
        pg = QPainter(self.thumb)
        linearGradient = QLinearGradient(0, 0, 0, 38)
        linearGradient.setColorAt(0.0, QColor(0x343434))
        linearGradient.setColorAt(0.5, QColor(0x606060))
        linearGradient.setColorAt(1.0, QColor(0x343434))
        pg.setBrush(QBrush(linearGradient))
        pg.drawRect(0, 2, self.thumb.width(), 36)

        p = QPainter(self.thumb)
        pen = QPen()
        col = QColor(0xc8c8c8)
        pen.setColor(col)
        p.setPen(pen)
        p.drawLine(0, 1, self.thumb.width(), 1)
        p.drawLine(0, self.thumb.height() - 2 - 10, self.thumb.width(), self.thumb.height() - 2 - 10)
        p.fillRect(0, 9, TIMELINE_TRACK_FRAME_MAX_WIDTH, 22, QColor(0xffffff))
        pen.setColor(QColor(0x343434))
        p.setPen(pen)
        p.drawLine(0, self.thumb.height() - 1 - 10, self.thumb.width(), self.thumb.height() - 1 - 10)
        p.drawLine(0, 0, 0, self.thumb.height())
        p.drawLine(0, 0, self.thumb.width(), 0)

        if self.thumbPixmap:
            B.drawPixmap(self.thumb, self.thumbPixmap, 1, (22 - self.thumbPixmap.height()) * .5 + 9)
        if self.getIdx() == 1:
            pass

    def setHoldFrameCount(self, count, dx=1):
        self.changeWidth = (count - self.holdFrameCount) * TIMELINE_TRACK_FRAME_MAX_WIDTH * dx
        self.resize(count * TIMELINE_TRACK_FRAME_MAX_WIDTH, self.height())
        self.thumb.resize(self.width(), self.thumb.height())
        self.rightButton.move(self.width() - self.rightButton.width() + 1, self.rightButton.y())
        self.holdFrameCount = count
        self.refSImage.holdFrameCount = count
        if count > 1:
            if not self.holdFrameCountLabel:
                self.holdFrameCountLabel = QLabel(self)
                self.holdFrameCountLabel.setAlignment(Qt.AlignRight)
                self.holdFrameCountLabel.setWordWrap(True)
                self.holdFrameCountLabel.setStyleSheet('color:#b6b6b6;border:none;')
                self.holdFrameCountLabel.setMinimumWidth(15)
                # self.holdFrameCountLabel.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Ignored)
                self.holdFrameCountLabel.show()
            self.holdFrameCountLabel.setText(str(count))
            self.holdFrameCountLabel.move(40 * count - 20, 10)
            print('holdFrameCountLabel', self.holdFrameCountLabel.text(), self.holdFrameCountLabel.rect(),
                  self.holdFrameCountLabel.x())

File: view/timeline/TrackFrame.cpp
//
// Created by manoyuria on 2015/6/19.
//

#include <model/App.h>
#include "TrackFrame.h"

TrackFrame::TrackFrame(QWidget *parent) : QWidget(parent) {
    resize(40, 60);

    thumb = new OverWidget<QWidget>(this);
    thumb->resize(40, 50);
    over(thumb, paintEvent_, paintThumb);
    _setMouseTransparent(thumb);


    frameIdx = new QLabel(this);
    frameIdx->setStyleSheet("color:#b6b6b6;border:none;");
    _setMouseTransparent(frameIdx);
    frameIdx->move(4, 45);
    frameIdx->setText("0");


    leftButton = new OverWidget<QPushButton>(this);
    leftButton->resize(8, 8);
    leftButton->isCustomQss = true;
    leftButton->add(mousePressEvent_, [this] { isPressLeftButton = true; });
    leftButton->add(mouseMoveEvent_, [this] { this->pressAndMoveEvent(); });
    leftButton->add(mouseReleaseEvent_, [this] { isPressLeftButton = false; });
    UI::setQss(leftButton, ":/qss_button", "frameButton");


    rightButton = new OverWidget<QPushButton>(this);
    rightButton->resize(8, 8);
    rightButton->move(33, 32);
    rightButton->isCustomQss = true;
    rightButton->add(mousePressEvent_, [this] { isPressRightButton = true; });
    rightButton->add(mouseMoveEvent_, [this] { this->pressAndMoveEvent(); });
    rightButton->add(mouseReleaseEvent_, [this] { isPressRightButton = false; });
    UI::setQss(rightButton, ":/qss_button", "frameButton");

    setMouseTracking(true);
}

void TrackFrame::paintThumb() {
    QPainter pg(this->thumb);
    QLinearGradient lg(0, 0, 0, 38);
    lg.setColorAt(0.0, QColor(0x343434));
    lg.setColorAt(0.5, QColor(0x606060));
    lg.setColorAt(1.0, QColor(0x343434));
    pg.setBrush(QBrush(lg));
    pg.drawRect(0, 2, width(), 36);

    QPainter p(this->thumb);
    QPen pen(QColor(0xc8c8c8));
    p.setPen(pen);
    p.drawLine(0, 1, this->thumb->width(), 1);
    p.drawLine(0, this->thumb->height() - 2 - 10, this->thumb->width(), this->thumb->height() - 2 - 10);
    p.fillRect(0, 9, TIMELINE_TRACK_FRAME_MAX_WIDTH, 22, QColor(0xffffff));
    pen.setColor(QColor(0x343434));
    p.setPen(pen);
    p.drawLine(0, this->thumb->height() - 1 - 10, this->thumb->width(), this->thumb->height() - 1 - 10);
    p.drawLine(0, 0, 0, this->thumb->height());
    p.drawLine(0, 0, this->thumb->width(), 0);

    QPainter pm(this->thumb);
    if (this->thumbPixmap != NULL) {
        pm.drawPixmap(int((TIMELINE_TRACK_FRAME_MAX_WIDTH- 2 - this->thumbPixmap->width() ) * .5),
                      int((22 - this->thumbPixmap->height()) * .5 + 9),
                      *this->thumbPixmap);
    }

}

void TrackFrame::setPixmap(QImage *qImage) {
    int thumbHeight = int(float(qImage->height()) / qImage->width() * (this->width() - 2));
    thumbPixmap = new QPixmap(QPixmap::fromImage(qImage->scaled(this->width() - 2, thumbHeight, Qt::KeepAspectRatio)));
}

void TrackFrame::pressAndMoveEvent() {
    if (isPressLeftButton || isPressRightButton) {
        int posX = mapFromGlobal(QCursor::pos()).x();
        if (isPressRightButton) {
            if (posX > width() + 30) {
                setHoldFrameCount(holdFrameCount + 1, 1);
                App()._().trackModel->relayout(trackInfoIdx,trackFrameInfoIdx);
            }
            else if (posX < width() - 30 and holdFrameCount > 1) {
                setHoldFrameCount(holdFrameCount - 1, 1);
                App()._().trackModel->relayout(trackInfoIdx,trackFrameInfoIdx);
            }
        }
        else if (isPressLeftButton) {
            if (posX < -30) {
                setHoldFrameCount(holdFrameCount + 1, -1);
                App()._().trackModel->relayout(trackInfoIdx,trackFrameInfoIdx);
            }
            else if (posX > 30 && holdFrameCount > 1) {
                setHoldFrameCount(holdFrameCount - 1, -1);
                App()._().trackModel->relayout(trackInfoIdx,trackFrameInfoIdx);
            }
        }
    }
}

void TrackFrame::setHoldFrameCount(int count, int dx = 1) {
    changeWidth = (count - holdFrameCount) * TIMELINE_TRACK_FRAME_MAX_WIDTH * dx;
    _setWidth(this, count * TIMELINE_TRACK_FRAME_MAX_WIDTH);
    _setWidth(thumb,width());
    _setX(rightButton, width() - rightButton->width() + 1);
    holdFrameCount = count;
}

void TrackFrame::setIdx(int trackFrameIdx) {
    idx = trackFrameIdx;
    QString s = QString::number(idx + 1, 10);
    frameIdx->setText(s);
}
